From a7b86c059d0d91ed9e0d9fe2326e1a13d5232627 Mon Sep 17 00:00:00 2001
From: Xavier Hanin <xavier@apache.org>
Date: Thu, 27 Apr 2006 07:02:11 +0000
Subject: [PATCH] FIX: Ivy file containing 2 different versions of the same
 module is not deliverd correctly (IVY-229) (thanks to Maarten Coene)

git-svn-id: https://svn.apache.org/repos/asf/incubator/ivy/trunk@484301 13f79535-47bb-0310-9956-ffa450edef68
---
 CHANGES.txt                                   |  1 +
 .../fr/jayasoft/ivy/ant/IvyDeliverTest.java   | 21 +++++++++++++++++++
 .../ivy/ant/ivy-different-revisions.xml       | 15 +++++++++++++
 3 files changed, 37 insertions(+)
 create mode 100644 test/java/fr/jayasoft/ivy/ant/ivy-different-revisions.xml

diff --git a/CHANGES.txt b/CHANGES.txt
index 2d4de2257..c98d09979 100644
--- a/CHANGES.txt
+++ b/CHANGES.txt
@@ -35,6 +35,7 @@ for detailed view of each issue, please consult http://jira.jayasoft.org/
   - IMPROVE: comments now aren't lost when delivering ivy files (IVY-226)
   - IMPROVE: add sort of 'fallback'-mapping to the defaultconfmapping attribute (IVY-215)
 
+  - FIX: Ivy file containing 2 different versions of the same module is not deliverd correctly (IVY-229)
   - FIX: ModuleRevisionId encode/decode doesn't work if revision is empty (IVY-233)
   - FIX: cachepath task should preserve the order of the dependencies (IVY-225)
   - FIX: specifying a defaultconfmapping adds dependencies of each unlisted configuration (IVY-214)
diff --git a/test/java/fr/jayasoft/ivy/ant/IvyDeliverTest.java b/test/java/fr/jayasoft/ivy/ant/IvyDeliverTest.java
index 82569c19c..8b2578c79 100644
--- a/test/java/fr/jayasoft/ivy/ant/IvyDeliverTest.java
+++ b/test/java/fr/jayasoft/ivy/ant/IvyDeliverTest.java
@@ -150,4 +150,25 @@ public void testNoReplaceDynamicRev() throws Exception {
         assertEquals(ModuleRevisionId.newInstance("org1", "mod1.2", "latest.integration"), dds[0].getDependencyRevisionId());
     }
 
+    public void testDifferentRevisionsForSameModule() throws Exception {
+        _project.setProperty("ivy.dep.file", "test/java/fr/jayasoft/ivy/ant/ivy-different-revisions.xml");
+        IvyResolve res = new IvyResolve();
+        res.setProject(_project);
+        res.execute();
+        
+        _deliver.setPubrevision("1.2");
+        _deliver.setDeliverpattern("build/test/deliver/ivy-[revision].xml");
+        _deliver.execute();
+        
+        // should have done the ivy delivering
+        File deliveredIvyFile = new File("build/test/deliver/ivy-1.2.xml");
+        assertTrue(deliveredIvyFile.exists()); 
+        ModuleDescriptor md = XmlModuleDescriptorParser.getInstance().parseDescriptor(new Ivy(), deliveredIvyFile.toURL(), true);
+        assertEquals(ModuleRevisionId.newInstance("jayasoft", "different-revs", "1.2"), md.getModuleRevisionId());
+        DependencyDescriptor[] dds = md.getDependencies();
+        assertEquals(3, dds.length);
+        assertEquals(ModuleRevisionId.newInstance("org1", "mod1.2", "2.0"), dds[0].getDependencyRevisionId());
+        assertEquals(ModuleRevisionId.newInstance("org1", "mod1.1", "1.0"), dds[1].getDependencyRevisionId());
+        assertEquals(ModuleRevisionId.newInstance("org1", "mod1.2", "1.1"), dds[2].getDependencyRevisionId());
+    }
 }
diff --git a/test/java/fr/jayasoft/ivy/ant/ivy-different-revisions.xml b/test/java/fr/jayasoft/ivy/ant/ivy-different-revisions.xml
new file mode 100644
index 000000000..9e4f3a137
--- /dev/null
+++ b/test/java/fr/jayasoft/ivy/ant/ivy-different-revisions.xml
@@ -0,0 +1,15 @@
+<ivy-module version="1.0"> 
+	<info organisation="jayasoft"
+	       module="different-revs"
+	       revision="1.0"
+	       status="release"
+	/>
+	<dependencies>
+		<dependency org="org1" name="mod1.2" rev="2.0"/>
+		<dependency org="org1" name="mod1.1" rev="1.0"/>
+		<dependency org="org1" name="mod1.2" rev="1.+"/>
+	</dependencies>
+	<conflicts>
+		<manager name="all" />
+	</conflicts>
+</ivy-module>
